x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/privileges/UserPrivilegeResolver.java

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+package org.elasticsearch.xpack.idp.privileges;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.client.Client;
+import org.elasticsearch.xpack.core.security.SecurityContext;
+import org.elasticsearch.xpack.core.security.action.user.HasPrivilegesAction;
+import org.elasticsearch.xpack.core.security.action.user.HasPrivilegesRequest;
+import org.elasticsearch.xpack.core.security.action.user.HasPrivilegesResponse;
+import org.elasticsearch.xpack.core.security.authz.RoleDescriptor;
+import org.elasticsearch.xpack.core.security.authz.permission.ResourcePrivileges;
+
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * Determines what privileges a user has within a given {@link ServiceProviderPrivileges service}.
+ */
+public class UserPrivilegeResolver {
+
+    public class UserPrivileges {
+        public final String principal;
+        public final boolean hasAccess;
+        public final Set<String> groups;
+
+        public UserPrivileges(String principal, boolean hasAccess, Set<String> groups) {
+            this.principal = principal;
+            this.hasAccess = hasAccess;
+            this.groups = groups;
+        }
+
+        @Override
+        public String toString() {
+            return getClass().getSimpleName() + "{" + principal + ", " + hasAccess + ", " + groups + "}";
+        }
+    }
+
+    private final Logger logger = LogManager.getLogger();
+    private final Client client;
+    private final SecurityContext securityContext;
+
+    public UserPrivilegeResolver(Client client, SecurityContext securityContext) {
+        this.client = client;
+        this.securityContext = securityContext;
+    }
+
+    /**
+     * Resolves the user's privileges for the specified service.
+     * Requires that the active user is set in the {@link org.elasticsearch.xpack.core.security.SecurityContext}.
+     */
+    public void resolve(ServiceProviderPrivileges service, ActionListener<UserPrivileges> listener) {
+        HasPrivilegesRequest request = new HasPrivilegesRequest();
+        final String username = securityContext.getUser().principal();
```

Comment:
> Requires that the active user is set in the {@link org.elasticsearch.xpack.core.security.SecurityContext}.

Should we also check here so that we don't throw an NPE on `.principal()` ?

Reply:
If you prefer a ISE over an NPE, then yeah, we can do that.

Reply:
It's not that much / only a personal preference, but I think we try to avoid NPEs throughout the codebase and check and fail with ISE / IAE when applicable so it felt more consistent. Just a nit though eitherway

Reply:
>  I think we try to avoid NPEs 

`Objects.requireNonNull` throws NPE, so I think we've accepted NPEs as long as the root cause is clear.